src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-13ubuntu/doc/gsmsmsstore.man
.\" -*- eval: (nroff-mode) -*-
.de TQ
.br
.ns
.TP \\$1
..
.\" Like TP, but if specified indent is more than half
.\" the current line-length - indent, use the default indent.
.de Tp
.ie \\n(.$=0:((0\\$1)*2u>(\\n(.lu-\\n(.iu)) .TP
.el .TP "\\$1"
..
.TH GSMSMSSTORE 8 "##DATE##" "gsmsmsstore v##VERSION##"
.SH NAME
gsmsmsstore \- SMS store manipulation program
.SH SYNOPSIS
.B gsmsmsstore
[ \fB\-a\fP ]
[ \fB\-\-add\fP ]
[ \fB\-b\fP \fIbaudrate\fP ]
[ \fB\-\-baudrate\fP \fIbaudrate\fP ]
[ \fB\-c\fP ]
[ \fB\-\-copy\fP ]
[ \fB\-C\fP \fIservice centre address\fP ]
[ \fB\-\-sca\fP \fIservice centre address\fP ]
[ \fB\-d\fP \fIdestination device or file\fP ]
[ \fB\-\-destination\fP \fIdestination device or file\fP ]
[ \fB\-h\fP ]
[ \fB\-\-help\fP ]
[ \fB\-I\fP \fIinit string\fP ]
[ \fB\-\-init\fP \fIinit string\fP ]
[ \fB\-k\fP ]
[ \fB\-\-backup\fP ]
[ \fB\-l\fP ]
[ \fB\-\-list\fP ]
[ \fB\-s\fP \fIsource device or file\fP ]
[ \fB\-\-source\fP \fIsource device or file\fP ]
[ \fB\-t\fP \fISMS store name\fP ]
[ \fB\-\-store\fP \fISMS store name\fP ]
[ \fB\-v\fP ]
[ \fB\-\-version\fP ]
[ \fB\-V\fP ]
[ \fB\-\-verbose\fP ]
[ \fB\-x\fP ]
[ \fB\-\-delete\fP ]
[ \fB\-X\fP ]
[ \fB\-\-xonxoff\fP ]
{ \fIindices\fP }
[ \fIphonenumber\fP \fItext\fP ]
.PP
.SH DESCRIPTION
\fIgsmsmsstore\fP can store or retrieve SMS messages entries residing
in a GSM mobile phone's SMS store to or from a file, add SMS messages
to a store, or list the store's contents. Additionally, it is possible
to add SMS submit messages to a store.
.PP
\fIgsmsmsstore\fP reads entries from the source which can be a mobile
phone (if a serial device file is given) or a file (if a file name is
given). The source is never modified. \fIgsmsmsstore\fP writes SMS
messages to a destination file or device in the case of \fB\-\-copy\fP,
\fB\-\-backup\fP, and \fB\-\-add\fP.
.PP
The \fB\-\-list\fP option does not change any file but just lists the
contents to standard output.
.PP
The \fB\-\-backup\fP and \fB\-\-copy\fP options require both source and
destination files or devices. The \fB\-\-list\fP option requires a
source. The \fB\-\-add\fP and \fB\-\-delete\fP options require a
destination file or device.
.PP
If "\-" is given as the parameter for the \fB\-\-source\fP or
\fB\-\-destination\fP options, the SMS store is read from standard input 
and/or written to standard output, respectively.
.PP
SMS message files are not human-readable.
.PP
Error messages are printed to the standard error output. If the program
terminates on error the error code 1 is returned.
.PP
.SH OPTIONS
.TP
\fB\-a\fP, \fB\-\-add\fP
Adds an SMS submit message with recipient address \fIphonenumber\fP and 
text \fItext\fP to the destination.
.TP
\fB\-b\fP \fIbaudrate\fP, \fB\-\-baudrate\fP \fIbaudrate\fP
The baud rate to use. The default baudrate is 38400.
.TP
\fB\-c\fP, \fB\-\-copy\fP
This causes the contents of the source to be copied to the
destination. After this operation the destination has exactly the same
contents as the source.  If \fIindices\fP are given on the command
line only those SMS messages denoted by the indices are copied to the
destination.
.TP
\fB\-C\fP \fIservice centre address\fP, \fB\-\-sca\fP \fIservice centre address\fP
Sets the service centre address to use for all SUBMIT SMSs (may not
work with some phones).
.TP
\fB\-d\fP \fIdestination\fP, \fB\-\-destination\fP \fIdestination\fP
The destination device or file.
.TP
\fB\-h\fP, \fB\-\-help\fP
Prints an option summary.
.TP
\fB\-I\fP \fIinit string\fP, \fB\-\-init\fP \fIinit string\fP
Initialization string to send to the TA (default: "E0"). Note that the
sequence "ATZ" is sent first.
.TP
\fB\-k\fP, \fB\-\-backup\fP
This causes those entries to be added from the source to the
destination that are not already present in the destination.  If
\fIindices\fP are given on the command line only those SMS messages
denoted by the indices are backed up (ie. added) to the destination.
.TP
\fB\-l\fP, \fB\-\-list\fP
Prints out the entire contents of the source in human-readable form.
.TP
\fB\-s\fP \fIsource\fP, \fB\-\-source\fP \fIsource\fP
The source device or file.
.TP
\fB\-t\fP \fISMS store name\fP, \fB\-\-store\fP \fISMS store name\fP
The name of the SMS store to read from or write to. This information is
only used for device sources and destinations. A commonly available message
store is "SM" (SIM card).
.TP
\fB\-v\fP, \fB\-\-version\fP
Prints the program version.
.TP
\fB\-V\fP, \fB\-\-verbose\fP
Prints out a detailed progress report.
.TP
\fB\-x\fP, \fB\-\-delete\fP
Delete the SMS messages as denoted by the \fIindices\fP from the destination.
.TP
\fB\-X\fP, \fB\-\-xonxoff\fP
Uses software handshaking (XON/XOFF) for accessing the device.
.PP
.SH EXAMPLES
The following command lists all entries in the mobile phone connected
to \fI/dev/mobilephone\fP to the standard output:
.PP
.nf
gsmsmsstore \-b 19200 \-s /dev/mobilephone \-t SM \-l
.fi
.PP
The following adds entries 4, 7, and 10 from the device
\fI/dev/mobilephone\fP to the file \fIsmsstore\fP:
.PP
.nf
gsmsmsstore \-s /dev/mobilephone \-d /home/fred/smsstore
    \-t SM \-b 4 7 10
.fi
.PP
.SH AUTHOR
Peter Hofmann <software@pxh.de>
.PP
.SH BUGS
Report bugs to software@pxh.de.  Include a complete, self-contained
example that will allow the bug to be reproduced, and say which
version of \fIgsmsmsstore\fP you are using.
.PP
.SH COPYRIGHT
Copyright \(co 1999 Peter Hofmann
.LP
\fIgsmsmsstore\fP is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public License as
published by the Free Software Foundation; either version 2, or (at
your option) any later version.
.LP
\fIgsmsmsstore\fP is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.
.LP
You should have received a copy of the GNU Library General Public License
along with \fIgsmsmsstore\fP; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.PP
.SH "SEE ALSO"
.BR gsminfo(7),
.BR gsmctl(1),
.BR gsmpb(1),
.BR gsmsendsms(1),
.BR gsmsmsd(8).


ChangeLog
 * libZRTP SDK library, implements the ZRTP secure VoIP protocol.
 * Copyright (c) 2006-2009 Philip R. Zimmermann.  All rights reserved.
 * Contact: http://philzimmermann.com
 * For licensing and other legal details, see the file zrtp_legal.c.

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
SINCE LIBZRTP v0.80 CHANGELOG IS A PART OF HTML DOCUMENTATION. 
https://developers.zfoneproject.com/libzrtp/wiki/LibzrtpChangeLog
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


libzrtp 0.7.1											18.11.2008
--------------------------------------------------------------------------------
1.	Fixed bug with hardcoded AES128 cipher for generating SRTP keys. In this 
	version is selected according to ZRTP discovery.
	
2.	Added initialization/deinitalization functions to zrtp helper functions.

libzrtp 0.7.0											04.11.2008
--------------------------------------------------------------------------------
1.	Changes in libzrtp sources tree.

2.	Improvements in libzrtp initialization routine:
	- all global zrtp options were combined in zrtp_config_t structure;
	- zrtp_init() allocates memory for zrtp global context;
	- zrtp_config_defaults() 

3.	Improvements for Passive/Active mode support.
	a) A passive endpoint never sends a commit message, period. Also, it 
	declares itself as a passive endpoint by setting the P flag it its own
	Hello message;
	b) A active endpoint does not send a commit to a passive endpoint, which it 
	recognizes by detecting the P flag;
	c) A passive phone, if acting as a SIP initiator (meaning it initiated the 
	call), rejects all commit packets from everyone;
	d) A passive phone rejects all commit messages from a PBX, which is easily 
	recognized by the M flag.
	Passive mode support is built into the library logic and will be used
	automatically if the developer specifies signaling role by setting 
	is_initiator flag in zrtp_init_session_ctx().

4.	Improvements in ZRTP feedback interface and system-dependent functions.
    There are two types of interface functions in libzrtp: system dependent API
	and realization of helper functions and events. System dependent API in
	defined in zrtp_iface_system.h as set of extern functions. System
	functions are already implemented for several basic platforms in
	zrtp_iface.sys.c libzrtp feedback and helper functions were re-factored
	and implemented as set of callbacks. If the developer doesn't want to
	handle one or another event it may just leave necessary pointer empty.
	See zrtp_init() and zrtp_callback_t, zrtp_iface.h for more details.

5. 	ZRTP configuration approach was improved: zrtp_config_xxx.h contains
	adjustments for necessary target platform. libzrtp contains default
	configs for Linux, OS X, Windows, Window CE and Symbian platforms.
	All ZRTP protocol and behavior related adjustments are collected in
	zrtp_config_user.h. Edit this file to configure libzrtp for your
	needs.
	
6.	Implemented new functions in protocol according to the Internet Draft v 10.

7.	Improved realization of built-in libzrtp scheduler. Fixed bug with crashing
	on performing delay call when zrtp session have been already deleted.
	
8. 	Logging function was improved. Use ZRTP_LOG macro to print log messages. See
	zrtp_log.h for more information.


libzrtp 0.6.8											03.09.2008
--------------------------------------------------------------------------------
ZRTP
1.	Fixed bug with incorrect maximum value for T1 retry interval timer which 
    increased delay between LOOKING_FOR_ZRTP and NO_ZRTP_SUPPORT states. 
    Internal fix, no API changes required;
2.	Implemented version negotiation according to the latest specification.
	libzrtp v 0.6.8 supports ZRTP v0.90 only. No changes required in 
	applications that use the SDK.
3. 	Compilation flag WITH_ZFONE was removed. The developer, who wants to use
	built-in ZRTP cache, has to set name of the ZRTP cache explicitly,
	implementing  zrtp_get_cache_path() function.
4. 	New libzrtp licensing scheme was implemented. It allows the licensing policy 
    to be changed at run time.  See zrtp_license_mode_t doc. for more details. 
    Affected API - zrtp_init().
5. 	ZRTP Protocol version was changed to 0.90 according to ZRTP Internet Draft.
6. 	Some changes in Linux config files: surplus configuration flags were removed
	from ./cfg.XXX templates. 
7.	Added experimental ZRTP messages retries scheduler for slow channels. As 
	ex ample for GSM CSD channel with average bandwidth 6Kb/s. To use this
	option build library with BUILD_FOR_CSD flag.


libzrtp 0.6.6											27.06.2008
--------------------------------------------------------------------------------
ZRTP
1. Small bug was fixed in S0 calculation: when RS1 is corrupted the library uses
   RS2 instead;
2. Some changes in Makefile and building process: unused header were eliminated 	
   from the installation process.
3. -DBUILD_WITH_ZRTP_MUTEXES was replaced with --enable_mutexes option passed to 
   ./configure script. This change allows not to specify any libzrtp compilation 
   flags during user application compilation. --enable-mutexes adds    
   BUILD_ZRTP_MUTEXES definition to the ./config/zrtp_unix_config.h so if you 
   build libzrtp on other platforms - define this flag manually (windows 
   configuration file already includes this option).
4. Clean-up in .h and .c comments was made.


libzrtp 0.6.5											04.06.2008
--------------------------------------------------------------------------------
ZRTP
1. New names for: other_secret - pbxs; srtps - auxs. In bits and secrets storages;
2. RS2 secret was eliminated form DH s0 calculation;
3. Protocol version number was increased to 0.85


libzrtp 0.6.4											19.05.2008
--------------------------------------------------------------------------------
ZRTP
1. According to the new version of the Internet Draft Signaling shared secret was
   removed from the protocol and from the sources. It was not used by interface

src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-12ubuntu1/apps/gsmsmsstore.cc
// *************************************************************************
// * GSM TA/ME library
// *
// * File:    gsmsmsstore.cc
// *
// * Purpose: SMS store management program
// *
// * Author:  Peter Hofmann (software@pxh.de)
// *
// * Created: 4.8.1999
// *************************************************************************

#ifdef HAVE_CONFIG_H
#include <gsm_config.h>
#endif
#include <gsmlib/gsm_nls.h>
#include <string>
#include <ctype.h>
#ifdef WIN32
#include <gsmlib/gsm_win32_serial.h>
#else
#include <gsmlib/gsm_unix_serial.h>
#include <unistd.h>
#endif
#if defined(HAVE_GETOPT_LONG) || defined(WIN32)
#include <getopt.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <gsmlib/gsm_me_ta.h>
#include <gsmlib/gsm_util.h>
#include <gsmlib/gsm_sorted_sms_store.h>
#include <iostream>

using namespace std;
using namespace gsmlib;

#ifdef HAVE_GETOPT_LONG
static struct option longOpts[] =
{
  {"xonxoff", no_argument, (int*)NULL, 'X'},
  {"init", required_argument, (int*)NULL, 'I'},
  {"store", required_argument, (int*)NULL, 't'},
  {"erase", no_argument, (int*)NULL, 'e'},
  {"add", no_argument, (int*)NULL, 'a'},
  {"list", no_argument, (int*)NULL, 'l'},
  {"destination", required_argument, (int*)NULL, 'd'},
  {"source", required_argument, (int*)NULL, 's'},
  {"baudrate", required_argument, (int*)NULL, 'b'},
  {"sca", required_argument, (int*)NULL, 'C'},
  {"copy", no_argument, (int*)NULL, 'c'},
  {"delete", no_argument, (int*)NULL, 'x'},
  {"backup", no_argument, (int*)NULL, 'k'},
  {"help", no_argument, (int*)NULL, 'h'},
  {"version", no_argument, (int*)NULL, 'v'},
  {"verbose", no_argument, (int*)NULL, 'V'},
  {(char*)NULL, 0, (int*)NULL, 0}
};
#else
#define getopt_long(argc, argv, options, longopts, indexptr) \
  getopt(argc, argv, options)
#endif

bool verbose = false;           // true if --verbose option given

// type of operation to perform

enum Operation {CopyOp = 'c', BackupOp = 'k', DeleteOp = 'x',
                AddOp = 'a', ListOp = 'l', NoOp = 0};

// aux function, insert entry only if not already present in dest

void backup(SortedSMSStoreRef destStore, SMSStoreEntry &entry)
{
  // the following only works because we know that the default sort order
  // is by date
  assert(destStore->sortOrder() == ByDate);

  Timestamp date = entry.message()->serviceCentreTimestamp();
  pair<SortedSMSStore::iterator, SortedSMSStore::iterator> range =
    destStore->equal_range(date);
          
  for (SortedSMSStore::iterator j = range.first;
       j != range.second; ++j)
    if (entry == *j)
      // do nothing if the entry is already present in the destination
      return;

  if (verbose)
    cout << stringPrintf(_("inserting entry #%d from source into destination"),
                         entry.index()) << endl
         << entry.message()->toString();
  destStore->insert(entry);     // insert
}

// aux function, throw exception if operation != NoOp

void checkNoOp(Operation operation, int opt)
{
  if (operation != NoOp)
    throw GsmException(stringPrintf(_("incompatible options '%c' and '%c'"),
                                    (char)operation, (char)opt),
                       ParameterError);
}

// *** main program

int main(int argc, char *argv[])
{
  try
  {
    // handle command line options
    string destination;
    string source;
    string baudrate;
    string storeName;
    char operation = NoOp;
    SortedSMSStoreRef sourceStore, destStore;
    bool useIndices = false;    // use indices in delete, copy, backup op
    string initString = DEFAULT_INIT_STRING;
    bool swHandshake = false;
    // service centre address (set on command line)
    string serviceCentreAddress;
    Ref<MeTa> sourceMeTa, destMeTa;

    int opt;
    int dummy;
    while((opt = getopt_long(argc, argv, "I:t:s:d:b:cxlakhvVXC:",
                             longOpts, &dummy))
          != -1)
      switch (opt)
      {
      case 'C':
        serviceCentreAddress = optarg;
        break;
      case 'X':
        swHandshake = true;
        break;
      case 'I':
        initString = optarg;
        break;
      case 'V':
        verbose = true;
        break;
      case 't':
        storeName = optarg;
        break;
      case 'd':
        destination = optarg;
        break;
      case 's':
        source = optarg;
        break;
      case 'b':
        baudrate = optarg;
        break;
      case 'c':
        checkNoOp((Operation)operation, opt);
        operation = CopyOp;
        break;
      case 'x':
        checkNoOp((Operation)operation, opt);
        operation = DeleteOp;
        break;
      case 'l':
        checkNoOp((Operation)operation, opt);
        operation = ListOp;
        break;
      case 'a':
        checkNoOp((Operation)operation, opt);
        operation = AddOp;
        break;
      case 'k':
        checkNoOp((Operation)operation, opt);
        operation = BackupOp;
        break;
      case 'v':
        cerr << argv[0] << stringPrintf(_(": version %s [compiled %s]"),
                                        VERSION, __DATE__) << endl;
        exit(0);
        break;
      case 'h':
        cerr << argv[0] << _(": [-a][-b baudrate][-c][-C sca]"
                             "[-d device or file]\n"
                             "  [-h][-I init string][-k][-l]"
                             "[-s device or file]"
                             "[-t SMS store name]\n  [-v][-V][-x][-X]"
                             "{indices}|[phonenumber text]") << endl
             << endl
             << _("  -a, --add         add new SMS submit message\n"
                  "                    (phonenumber and text) to destination")
             << endl
             << _("  -b, --baudrate    baudrate to use for device "
                  "(default: 38400)")
             << endl
             << _("  -c, --copy        copy source entries to destination\n"
                  "                    (if indices are given, "
                  "copy only these entries)") << endl
             << _("  -C, --sca         SMS service centre address") << endl
             << _("  -d, --destination sets the destination device to\n"
                  "                    connect to, or the file to write to")
             << endl
             << _("  -h, --help        prints this message") << endl
             << _("  -I, --init        device AT init sequence") << endl
             << _("  -k, --backup      backup new entries to destination\n"
                  "                    (if indices are given, "
                  "copy only these entries)") << endl
             << _("  -l, --list        list source to stdout") << endl
             << _("  -s, --source      sets the source device to connect to,\n"
                  "                    or the file to read") << endl
             << _("  -t, --store       name of SMS store to use") << endl
             << _("  -v, --version     prints version and exits") << endl
             << _("  -V, --verbose     print detailed progress messages")
             << endl
             << _("  -x, --delete      delete entries denoted by indices")
             << endl
             << _("  -X, --xonxoff     switch on software handshake") << endl
             << endl;
        exit(0);
        break;
      case '?':
        throw GsmException(_("unknown option"), ParameterError);
        break;
      }

    // check if parameters are complete
    if (operation == NoOp)
      throw GsmException(_("no operation option given"), ParameterError);
    if (operation == BackupOp || operation == CopyOp)
      if (destination.length() == 0 || source.length() == 0)
        throw GsmException(_("both source and destination required"),
                           ParameterError);
    if (operation == ListOp)
    {
      if (destination.length() != 0)
        throw GsmException(_("destination must not be given"), ParameterError);
      if (source.length() == 0)
        throw GsmException(_("source required"), ParameterError);
    }
    if (operation == AddOp || operation == DeleteOp)
    {
      if (source.length() != 0)
        throw GsmException(_("source must not be given"), ParameterError);
      if (destination.length() == 0)
        throw GsmException(_("destination required"), ParameterError);
    }
    if (operation == CopyOp || operation == DeleteOp || operation == BackupOp)
    {
      // check if all indices are numbers
      for (int i = optind; i < argc; ++i)
        for (char *pp = argv[i]; *pp != 0; ++pp)
          if (! isdigit(*pp))
            throw GsmException(stringPrintf(_("expected number, got '%s'"),
                                            argv[i]), ParameterError);
      useIndices = optind != argc;
    }
    else if (operation == AddOp)
    {
      if (optind + 2 < argc)
        throw GsmException(_("more than two parameters given"),
                           ParameterError);
      if (optind + 2 > argc)
        throw GsmException(_("not enough parameters given"),
                           ParameterError);
    }
    else
      if (optind != argc)
        throw GsmException(_("unexpected parameters"), ParameterError);
    
    // start accessing source store or file if required by operation
    if (operation == CopyOp || operation == BackupOp || operation == ListOp)
      if (source == "-")
        sourceStore = new SortedSMSStore(true);
      else if (isFile(source))
        sourceStore = new SortedSMSStore(source);
      else
      {
        if (storeName == "")
          throw GsmException(_("store name must be given"), ParameterError);
        
        sourceMeTa = new MeTa(new
#ifdef WIN32
                              Win32SerialPort
#else
                              UnixSerialPort
#endif
                              (source,
                               baudrate == "" ? DEFAULT_BAUD_RATE :
                               baudRateStrToSpeed(baudrate), initString,
                               swHandshake));
        sourceStore = new SortedSMSStore(sourceMeTa->getSMSStore(storeName));
      }
      
    // make sure destination file exists
    if (destination != "")
    {
      try
      {
        ofstream f(destination.c_str(), ios::out | ios::app | ios::binary);
      }
      catch (exception)
      {
      }
    }

    // start accessing destination destination store or file
    if (operation == CopyOp || operation == BackupOp || operation == AddOp ||
        operation == DeleteOp)
      if (destination == "-")
        destStore = new SortedSMSStore(false);
      else if (isFile(destination))
        destStore = new SortedSMSStore(destination);
      else
      {
        if (storeName == "")
          throw GsmException(_("store name must be given"), ParameterError);
        
        destMeTa = new MeTa(new
#ifdef WIN32
                            Win32SerialPort
#else
                            UnixSerialPort
#endif
                            (destination,
                             baudrate == "" ? DEFAULT_BAUD_RATE :
                             baudRateStrToSpeed(baudrate), initString,
                             swHandshake));
        destStore = new SortedSMSStore(destMeTa->getSMSStore(storeName));      
      }

    // now do the actual work
    switch (operation)
    {
    case BackupOp:
    {
      sourceStore->setSortOrder(ByIndex); // needed in loop

      if (useIndices)
        for (int i = optind; i < argc; ++i)
        {
          SortedSMSStore::iterator j = sourceStore->find(atoi(argv[i]));
          if (j == sourceStore->end())
            throw GsmException(stringPrintf(_("no index '%s' in source"),
                                            argv[i]), ParameterError);
          backup(destStore, *j);
        }
      else
        for (SortedSMSStore::iterator i = sourceStore->begin();
             i != sourceStore->end(); ++i)
          backup(destStore, *i);
      break;
    }
    case CopyOp:
    {                        
      destStore->clear();
      if (! useIndices)         // copy all entries
      {
        for (SortedSMSStore::iterator i = sourceStore->begin();
             i != sourceStore->end(); ++i)
        {
          if (verbose)
            cout << stringPrintf(_("inserting entry #%d from source "
                                   "into destination"), i->index()) << endl
                 << i->message()->toString();
          destStore->insert(*i);
        }
      }
      else                      // copy indexed entries
      {
        sourceStore->setSortOrder(ByIndex); // needed in loop

        for (int i = optind; i < argc; ++i)
        {
          SortedSMSStore::iterator j = sourceStore->find(atoi(argv[i]));
          if (j == sourceStore->end())
            throw GsmException(stringPrintf(_("no index '%s' in source"),
                                            argv[i]), ParameterError);
          if (verbose)
            cout << stringPrintf(_("inserting entry #%d from source into "
                                   "destination"), j->index()) << endl
                 << j->message()->toString();
          destStore->insert(*j);
        }
      }
      break;
    }
    case ListOp:
    {
      for (SortedSMSStore::iterator i = sourceStore->begin();
           i != sourceStore->end(); ++i)
        cout << stringPrintf(_("index #%d"), i->index()) << endl
             << i->message()->toString();
      break;
    }
    case AddOp:
    {
      SMSMessageRef sms = new SMSSubmitMessage(argv[optind + 1], argv[optind]);
      // set service centre address in new submit PDU if requested by user
      if (serviceCentreAddress != "")
      {
        Address sca(serviceCentreAddress);
        sms->setServiceCentreAddress(sca);
      }
      if (verbose)
        cout << _("inserting new entry into destination") << endl
             << sms->toString();
      destStore->insert(sms);
      break;
    }
    case DeleteOp:
    {
      destStore->setSortOrder(ByIndex);
      for (int i = optind; i < argc; ++i)
      {
        int index = atoi(argv[i]);
        if (verbose)
        {
          SortedSMSStore::iterator e = destStore->find(index);
          if (e != destStore->end())
            cout << stringPrintf(_("deleting entry #%d from destination"),
                                 index) << endl
                 << e->message()->toString();
        }
        if (destStore->erase(index) != 1)
          throw GsmException(stringPrintf(_("no index '%s' in destination"),
                                          argv[i]), ParameterError);
      }
      break;
    }
    }
  }
  catch (GsmException &ge)
  {
    cerr << argv[0] << _("[ERROR]: ") << ge.what() << endl;
    return 1;
  }
  return 0;
}
